fontir/src/orchestration.rs

```diff
@@ -0,0 +1,108 @@
+// Helps coordinate the graph execution
+// Eventually has to move outside fontir to let us use for BE work
+
+use std::{
+    collections::{HashMap, HashSet},
+    sync::Arc,
+};
+
+use parking_lot::RwLock;
+
+use crate::ir::{GlobalMetadata, GlyphIr};
+
+// Unique identifier of work. If there are no fields work is unique.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+pub enum WorkIdentifier {
+    GlobalMetadata,
+    GlyphIr(u32),
+    FinishIr,
+}
+
+const MISSING_DATA: &str = "Missing data, dependency management failed us?";
+
+/// Read/write access to data for async work.
+///
+/// Intent is a root orchestrator creates a context and share copies with restricted
+/// access with spawned tasks. Copies with access control are created to detect bad
+/// execution order / mistakes, not to block actual bad actors.
+pub struct Context {
+    // If present, the one and only key you are allowed to write to
+    // Otherwise you totally get to write whatever you like
+    write_mask: Option<WorkIdentifier>,
+
+    // If present, what you can access through this context
+    // Intent is root has None, task-specific Context only allows access to dependencies
+    read_mask: Option<HashSet<WorkIdentifier>>,
+
+    // work results we've completed or restored from disk
+    // We create individual caches so we can return typed results from get fns
+    global_metadata_cache: Arc<RwLock<Option<Arc<GlobalMetadata>>>>,
+    glyph_ir_cache: Arc<RwLock<HashMap<u32, Arc<GlyphIr>>>>,
+}
+
+impl Context {
+    pub fn new_root() -> Context {
+        Context {
+            write_mask: None,
+            read_mask: None,
+            global_metadata_cache: Arc::from(RwLock::new(None)),
+            glyph_ir_cache: Arc::from(RwLock::new(HashMap::new())),
+        }
+    }
+}
+
+impl Context {
+    pub fn copy_for_work(
+        &self,
+        work_id: WorkIdentifier,
+        dependencies: HashSet<WorkIdentifier>,
+    ) -> Context {
+        Context {
+            write_mask: Some(work_id),
+            read_mask: Some(dependencies),
+            global_metadata_cache: self.global_metadata_cache.clone(),
+            glyph_ir_cache: self.glyph_ir_cache.clone(),
+        }
+    }
+
+    fn check_read_access(&self, id: &WorkIdentifier) {
+        if !self
+            .read_mask
+            .as_ref()
+            .map(|mask| mask.contains(id))
+            .unwrap_or(true)
+        {
+            panic!("Illegal access");
+        }
+    }
+
+    fn check_write_access(&self, id: &WorkIdentifier) {
+        if !self.write_mask.map(|mask| mask == *id).unwrap_or(true) {
+            panic!("Illegal access");
+        }
+    }
```

Comment:
So something to consider here: you are describing an invariant ("type must have access in order to perform task") and enforcing it through a convention ("you have to call this method anytime you want to read or write") but you could also consider enforcing it *through the type system itself*.

e.g, imagine the following:

```rust
struct ReadableContext<'a>(&'a Context);

impl Context {
    fn readable(&'a self, id: &WorkIdentifier) -> ReadableContext<'a> {
        if !self.can_read(id) {
            panic!("Illegal access");
        }
        ReadableContext(self)
    }
}
```

(imagine something similar for `WriteableContext`)

And then all the methods for reading/writing exist only on the appropriate contexts, so you actually can't call them at all unless you've gone through this check method.

I'm not sure if it's appropriate here, but in general this is a nice pattern for ensuring invariants are upheld, and it becomes more and more valuable as a project gets more complex.

Reply:
The problem I ran into with such a model is that any given context can read some arbitrary subset of readable things and that doesn't express in types very well as far as I can tell. For example, imagine a BE glyph task whose dependencies are {global metadata, IR for the glyph and it's components} ... how do I model that in the type system?

Reply:
Yea, it may not be the right fit in this instance, but wanted to at least bring it up as a design pattern I would consider. :)

Reply:
It's a glorious model, I just couldn't quite see how to make it work.